config/v1/types_image_digest_mirror_set.go

```diff
@@ -0,0 +1,113 @@
+}
+
+// ImageDigestMirrors holds cluster-wide information about how to handle mirrors in the registries config.
+type ImageDigestMirrors struct {
+	// source matches the repository that users refer to, e.g. in image pull specifications.
+	// "source" uses one of the following formats:
+	// host[:port]
+	// host[:port]/namespace[/namespace…]
+	// host[:port]/namespace[/namespace…]/repo
+	// [*.]host
+	// for more information about the format, see the document about the location field:
+	// https://github.com/containers/image/blob/main/docs/containers-registries.conf.5.md#choosing-a-registry-toml-table
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^\*(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+$|^((?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?)(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?$`
+	Source string `json:"source"`
+	// mirrors is zero or more locations that may also contain the same images.
+	// Images can be pulled from these mirrors only if they are referenced by their digests.
+	// The mirrored location is obtained by replacing the part of the input reference that
+	// matches source by the mirrors entry, e.g. for registry.redhat.io/product/repo reference,
+	// a (source, mirror) pair *.redhat.io, mirror.local/redhat causes a mirror.local/redhat/product/repo
+	// repository to be used.
+	// If the "mirrors" is not specified, the image will continue to be pulled from the specified
+	// repository in the pull spec. No mirror will be configured.
+	// The order of mirrors in this list is treated as the user's desired priority, while source
+	// is by default considered lower priority than all mirrors. Other cluster configuration,
+	// including (but not limited to) other imageDigestMirrors objects,
+	// may impact the exact order mirrors are contacted in, or some mirrors may be contacted
+	// in parallel, so this should be considered a preference rather than a guarantee of ordering.
+	// "mirrors" uses one of the following formats:
+	// host[:port]
+	// host[:port]/namespace[/namespace…]
+	// host[:port]/namespace[/namespace…]/repo
+	// for more information about the format, see the document about the location field:
+	// https://github.com/containers/image/blob/main/docs/containers-registries.conf.5.md#choosing-a-registry-toml-table
+	// +optional
+	// +listType=set
+	Mirrors []Mirrors `json:"mirrors,omitempty"`
+	// When enabled, prevents image pull from the specified repository in the pull spec if the image pull form the mirror list fails.
+	// Default is false, the image will continue to be pulled from the pull spec that matches the "source" if the image can not be pulled from the mirror list.
+	// neverContactSource is valid configuration only when one or more mirrors are in the mirror list.
+	// +optional
+	NeverContactSource bool `json:"neverContactSource,omitempty"`
```

Comment:
we don't use bools in the API, we describe policies using enums.  In this case, I would expect something like, `MirrorPullPolicy` with values "SourceFirst", "SourceLast", "SourceNever", "", where "" means no opinion and the platform chooses a default which is subject to change.

Reply:
If we cannot use bool in the API, I propose we replace the NeverContactSource with the following:
@umohnani8?
```go	

type SourcePolicy string

const (
	NeverContactSource SourcePolicy = "sourceNever"
	LastContactSource  SourcePolicy = "sourceLast"
)

SourcePolicy       SourcePolicy `json:"sourcePolicy,omitempty"`
```

Reply:
`LastContactSource` is not accurate.

The behavior of the code is that we _allow_ including `Source` (if it is not a hostname wildcard pattern) as a member of `Mirrors`; in that case, `Source` is contacted in the order it is listed in `Mirrors` (modulo the concerns about merging multiple CRs with the same `Source`). Only if `Source` is not listed in `Mirrors`, it is implicitly appended at the end.

The effective behavior we are proposing for this tunable is:
- Default: `Mirrors` are contacted, generally in the order they are listed; if `Source` is not included in the `Mirrors` list, it is implicitly contacted as the last resort.
- `neverContactSource`: The same, but access to (the physical registry) `Source` will always immediately fail; effectively `Source` is never contacted, only the other mirrors are.

So a better fit would probably be a set of `neverContactSource`/`allowContactingSource`. (I can’t imagine a third possible option, but it’s quite likely one will come up eventually.)

---

That leads to an interesting corner case: We should probably reject, or at least warn about, a configuration where `neverContactSource` is used, while `Source` is included as an element of `Mirrors`.